### PR TITLE
Staylor/demos

### PR DESCRIPTION
**Pull request by staylorTT**
Adding more demos to the repo.

### CODE REVIEW

**Review by nsmithtt — APPROVED** [review]

File: demos/tt-forge-fe/README.md [hunk 1]
```diff
@@ -21,14 +21,30 @@ Before running the demos, make sure you have TT-Forge-FE installed. You can inst
...
-python cnn/resnet_50_demo.py
+# CNN Models
+python cnn/resnet_50_demo.py              # Run ResNet-50 demo
+python cnn/mobile_netv2_demo.py           # Run MobileNetV2 demo
+python cnn/vilt_demo.py                   # Run ViLT demo
+python cnn/efficientnet_timm_demo.py      # Run EfficientNet (timm) demo
+
+# NLP Models
+python nlp/bert_demo.py                   # Run BERT demo
+python nlp/bert_demo_turkish.py           # Run Turkish BERT demo
```
Comment (nvukobratTT) [inline demos/tt-forge-fe/README.md:44]: Turkish BERT, too specific lol 🇹🇷

File: demos/tt-forge-fe/cnn/efficientnet_timm_demo.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import torch
+import timm
+from PIL import Image
+import requests
+from torchvision import transforms
+from timm.data import resolve_data_config
+from timm.data.transforms_factory import create_transform
+import forge
+
+class EfficientNetWrapper(torch.nn.Module):
```
Comment (nvukobratTT) [inline demos/tt-forge-fe/cnn/efficientnet_timm_demo.py:13]: This wrapper is probably redundant. Config can be set outside of it, and the rest of the code is just calling regular inference of the model without any modifications. If we want cleaner demo, we should remove it.

@staylorTT did you spot any specific reason for keeping it?

File: demos/tt-forge-fe/cnn/efficientnet_timm_demo.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    def __init__(self, model):
+        super().__init__()
+        self.model = model
+        self.config = resolve_data_config({}, model=model)
+
+    def forward(self, x):
+        return self.model(x)
+
+def run_efficientnet_inference(image_url="https://raw.githubusercontent.com/pytorch/hub/master/images/dog.jpg"):
```
Comment (nvukobratTT) [inline demos/tt-forge-fe/cnn/efficientnet_timm_demo.py:22]: Let's just keep image URL in the __main__ part of the function. Just too have a cleaner main inference function call.

File: demos/tt-forge-fe/cnn/efficientnet_timm_demo.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    # Create input sample for compilation
+    input_sample = (tensor,)
+
+    print("Compiling model...")
+    compiled_model = forge.compile(wrapper, input_sample)
+
+    # Run inference
+    print("Running inference...")
+    with torch.no_grad():
```
Comment (nvukobratTT) [inline demos/tt-forge-fe/cnn/efficientnet_timm_demo.py:45]: This one is redundant as well. As model is in eval model (model.eval()), gradients are already set to false.

File: demos/tt-forge-fe/cnn/efficientnet_timm_demo.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        output = compiled_model(*input_sample)
+    
+    # Get predictions
+    logits = output[0]  # Get first output tensor
+    print("Logits shape:", logits.shape)
```
Comment (nvukobratTT) [inline demos/tt-forge-fe/cnn/efficientnet_timm_demo.py:50]: Do we need thise debug prints for demo?

File: demos/tt-forge-fe/cnn/efficientnet_torchvision_demo.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+from PIL import Image
+import requests
+from torchvision import transforms
+import forge
+
+class EfficientNetWrapper(torch.nn.Module):
+    def __init__(self, model):
+        super().__init__()
+        self.model = model
+        self.transform = transforms.Compose([
```
Comment (nvukobratTT) [inline demos/tt-forge-fe/cnn/efficientnet_torchvision_demo.py:15]: We should have a unified way of how our demo script looks. E.g. where we're doing input pre-post processing, how we're loading the model, etc. For example, here you defined pre-processing as part of wrapper, while test above has it as part of demo script. To enable users to easier follow our demos, we should have a more precise structure.

File: demos/tt-forge-fe/cnn/efficientnet_torchvision_demo.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            transforms.Resize(256),
+            transforms.CenterCrop(224),
+            transforms.ToTensor(),
+            transforms.Normalize(
+                mean=[0.485, 0.456, 0.406],
+                std=[0.229, 0.224, 0.225]
+            )
+        ])
+
+    def forward(self, x):
```
Comment (nvukobratTT) [inline demos/tt-forge-fe/cnn/efficientnet_torchvision_demo.py:25]: Also, I think this wrapper is redundant as well

File: demos/tt-forge-fe/cnn/efficientnet_torchvision_demo.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        return self.model(x)
+
+def run_efficientnet_inference(image_url="https://raw.githubusercontent.com/pytorch/hub/master/images/dog.jpg"):
+    # Create model
+    model = models.efficientnet_b0()
+    # Download weights directly from torchvision
+    state_dict = torch.hub.load_state_dict_from_url(
```
Comment (nvukobratTT) [inline demos/tt-forge-fe/cnn/efficientnet_torchvision_demo.py:32]: You should be able to load weights in much more cleaner way. E.g.
```
resnet50(weights=ResNet50_Weights.IMAGENET1K_V2)
```

Checkout more details here:
- https://pytorch.org/vision/0.20/models.html#initializing-pre-trained-models
- https://pytorch.org/vision/main/models/generated/torchvision.models.efficientnet_b0.html#torchvision.models.EfficientNet_B0_Weights

File: demos/tt-forge-fe/cnn/efficientnet_torchvision_demo.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+    # Create input sample for compilation
+    input_sample = (tensor,)
+
+    print("Compiling model...")
+    compiled_model = forge.compile(wrapper, input_sample)
+
+    # Run inference
+    print("Running inference...")
+    with torch.no_grad():
```
Comment (nvukobratTT) [inline demos/tt-forge-fe/cnn/efficientnet_torchvision_demo.py:55]: Also redundant

File: demos/tt-forge-fe/cnn/vilt_demo.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
+# SPDX-FileCopyrightText: 2025 Tenstorrent AI ULC
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+import requests
+from PIL import Image
+from transformers import ViltProcessor, ViltForQuestionAnswering, ViltConfig
+import forge
+
+class ViltEmbeddingWrapper(torch.nn.Module):
```
Comment (nvukobratTT) [inline demos/tt-forge-fe/cnn/vilt_demo.py:10]: Vilt is a mess.. We (dev team) should revisit this one it these wrappers are really required in order to run model e2e

File: demos/tt-forge-fe/cnn/vilt_demo.py [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+
+        if self.task == "qa":
+            logits = self.vilt_model.classifier(pooled_output)
+            viltmodel_output = (logits,) + viltmodel_output[2:]
+
+        return viltmodel_output
+
+def load_image_from_url(url):
+    """Load an image from a URL with caching."""
+    import os
```
Comment (nvukobratTT) [inline demos/tt-forge-fe/cnn/vilt_demo.py:75]: Let's avoid imports within the functions

File: demos/tt-forge-fe/cnn/vilt_demo.py [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+    import hashlib
+    
+    # Create cache directory if it doesn't exist
```
Comment (nvukobratTT) [inline demos/tt-forge-fe/cnn/vilt_demo.py:78]: Can't we just load the image in a similar way as in other examples? 

This is a bit complicated way of loading up the image, and probably not required

**Review by nvukobratTT — COMMENTED** [review]
Let a few comments, but they apply to other scripts as well. 

General comment, when we start with per push we'll probably want to run pre and post processing on our device as well. This means capturing everything in model wrapper.

**Comment by nvukobratTT** [comment]
@vkovinicTT @mstojkovicTT @dgolubovicTT I left few comments on this PR, feel free to leave more if needed. 

Also, as I'm on vacation, please help out @staylorTT next week to correct these demos as much as possible. Just have in mind that we don't need to fix complicated issues right now. For anything that requires compiler changes, let's correct them latter and update these demos. For now, I don't want to block this PR to be merged if it needs to wait compiler fixes :))

File: demos/tt-forge-fe/cnn/vilt_demo.py (hunk 10 above)
Comment (staylorTT) [reply demos/tt-forge-fe/cnn/vilt_demo.py:10]: This one took windsurf a long time to figure out lol. Once I pointed it to the unit test it was able to get something running. Happy to have a dev clean it up however it makes sense.

File: demos/tt-forge-fe/llm/falcon3_demo.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        print(f"\nPrompt: {prompt}")
+        
+        # Tokenize input
+        inputs = tokenizer(prompt, return_tensors="pt")
+        input_ids = inputs['input_ids']
+        attention_mask = inputs['attention_mask']
+        
+        # Generate response using original model since generation requires the full model
+        with torch.no_grad():
+            output = model.generate(
```
Comment (dgolubovicTT) [inline demos/tt-forge-fe/llm/falcon3_demo.py:59]: In this demo we don't use compiled_model... Does inference work for falcon3? If not, why the demo?

**Review by dgolubovicTT — CHANGES_REQUESTED** [review]
Checkout falcon3 test. Why do we need it, if we don't use compiled model.
